Python/pytime.c

```diff
@@ -42,6 +42,27 @@ _PyTime_overflow(void)
                     "timestamp too large to convert to C _PyTime_t");
 }
 
+
+#if defined(MS_WINDOWS) || defined(__APPLE__)
+Py_LOCAL_INLINE(_PyTime_t)
+_PyTime_MulDiv(_PyTime_t ticks, _PyTime_t mul, _PyTime_t div)
+{
+    _PyTime_t sec, nsec;
+    /* Compute ticks * mul / div in two parts to prevent integer overflow:
+       compute the number of seconds ("integer part"), and then the remaining
+       number of nanoseconds ("floating part").
+
+       The caller has to check that ticks * mul, with ticks < div, cannot
+       overflow. */
+    sec = ticks / div;
+    ticks %= div;
+    nsec = ticks * mul;
+    nsec /= div;
+    return sec * mul + nsec;
```

Comment:
I'm wondering if it might be clearer to spell out the full calculations, and then rely on the compiler being smart enough to figure out whether or not there are any subcalculations it can re-use:

    sec = ticks / div * mul; /* Avoid overflow in the integer part */
    nsec = (ticks % div) * mul / div; /* Avoid loss of precision in the fractional part */
    return sec + nsec

If you write it out that way, then the suggested comment above could be simplified to:

> The caller must ensure that "div * mul" cannot overflow, as this ensures that neither the seconds calculation nor the nanoseconds calculation will overflow.

Reply:
I prefer to leave the code as it is. I prefer to write explicit code and don't rely on the computer to be smart.

I just hope that some smart compilers are able to compile "ticks / div" and "ticks % div" as a single CPU instruction ;-) (At least, Intel AMD64 is able to do that.)

Reply:
Nick's shorter variant looks clearer to me too.

But the name `sec` looks misleading. This isn't a time in seconds, and if it would be, adding seconds and nanoseconds would be wrong.

Reply:
That's a good point about the names: `sec_part` and `nsec_part` would be more accurate (since they're both in units of nanoseconds, it's just that the first one has a zero nanosecond component, while the latter has a zero seconds component).

Regarding "what are compilers and CPUs good at optimising?", the thing you really, really, really want to avoid is pipeline stalls as they wait to touch the same piece of memory multiple times without enough other work in between. Setting out the calculations as distinct operations makes it clear to the compiler that the two calculations are entirely independent of each other, and the only time they need to be synchronised is in the final addition.

So that's why I favour the approach of making the algorithmic *intent* as clear as possible, for the benefit of both human readers *and* compilers, and in this case, that means making the two expressions independent of each other rather than requiring the reader to track the local variable mutation and reverse engineer the original expressions from that.

Hand-optimisation *might* make sense if a profiler indicated this was a hotspot, but even then, I'd be surprised if switching from "two independent subexpressions" to "touch the same local variable multiples times" actually sped things up.

Reply:
I renamed variables to intpart and remaining since I'm trying to not hardcode the _PyTime_t unit in pytime.c. The API is "unit agnostic".